dev/src/rate-limiter.ts

```diff
@@ -0,0 +1,150 @@
+ *
+ * RateLimiter can also implement a gradually increasing rate limit. This is
+ * used to enforce the 500/50/5 rule
+ * (https://cloud.google.com/datastore/docs/best-practices#ramping_up_traffic).
+ *
+ * @private
+ */
+export class RateLimiter {
+  // Number of tokens available. Each operation consumes one token.
+  availableTokens: number;
+
+  // When the token bucket was last refilled.
+  lastRefillTimeMillis: number;
+
+  /**
+   * @param initialCapacity Initial maximum number of operations per second.
+   * @param multiplier Rate by which to increase the capacity.
+   * @param multiplierMillis How often the capacity should increase in
+   * milliseconds.
+   * @param startTimeMillis The starting time in epoch milliseconds that the rate limit is based on. Used
+   * for testing the limiter.
+   */
+  constructor(
+    private readonly initialCapacity: number,
+    private readonly multiplier: number,
+    private readonly multiplierMillis: number,
+    private readonly startTimeMillis = Date.now()
+  ) {
+    this.availableTokens = initialCapacity;
+    this.lastRefillTimeMillis = startTimeMillis;
+  }
+
+  /**
+   * Tries to make the number of operations. Returns true if the request
+   * succeeded and false otherwise.
+   *
+   * @param currentTimeMillis The date used to calculate the number of available
+   * tokens. Used for testing the limiter.
+   * @private
+   */
+  tryMakeRequest(
+    numOperations: number,
+    currentTimeMillis = Date.now()
+  ): boolean {
+    this.refillTokens(currentTimeMillis);
+    if (numOperations <= this.availableTokens) {
+      this.availableTokens -= numOperations;
+      return true;
+    }
+    return false;
+  }
+
+  /**
+   * Returns the number of ms needed to make a request with the provided number
+   * of operations. Returns 0 if the request can be made with the existing
+   * capacity. Returns -1 if the request is not possible with the current
+   * capacity.
+   *
+   * @param currentTimeMillis The date used to calculate the number of available
+   * tokens. Used for testing the limiter.
+   * @private
+   */
+
+  getNextRequestDelayMs(
+    numOperations: number,
+    currentTimeMillis = Date.now()
+  ): number {
+    if (numOperations < this.availableTokens) {
+      return 0;
+    }
+
+    const capacity = this.calculateCapacity(currentTimeMillis);
+    if (capacity < numOperations) {
+      return -1;
+    }
+
+    const requiredTokens = numOperations - this.availableTokens;
+    return Math.ceil((requiredTokens * 1000) / capacity);
+  }
+
+  /**
+   * Refills the number of available tokens based on how much time has elapsed
+   * since the last time the tokens were refilled.
+   *
+   * @param currentTimeMillis The date used to calculate the number of available
+   * tokens. Used for testing the limiter.
+   * @private
+   */
+  private refillTokens(currentTimeMillis = Date.now()): void {
+    if (currentTimeMillis > this.lastRefillTimeMillis) {
```

Comment:
We might want to make the behavior of this class more predictable and reject timestamps that go back in time.

Reply:
Added behavior to reject timestamps that are before the last refill time. Also added test for this.